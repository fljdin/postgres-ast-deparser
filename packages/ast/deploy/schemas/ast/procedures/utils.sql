-- Deploy schemas/ast/procedures/utils to pg

-- requires: schemas/ast/schema

BEGIN;


COMMIT;
