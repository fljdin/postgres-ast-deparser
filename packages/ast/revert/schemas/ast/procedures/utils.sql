-- Revert schemas/ast/procedures/utils from pg

BEGIN;

DROP FUNCTION ast.utils;

COMMIT;
