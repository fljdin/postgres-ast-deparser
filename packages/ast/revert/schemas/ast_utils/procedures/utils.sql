-- Revert schemas/ast_utils/procedures/utils from pg

BEGIN;

DROP FUNCTION ast_utils.utils;

COMMIT;
