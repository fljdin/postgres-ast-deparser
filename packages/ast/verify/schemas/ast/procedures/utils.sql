-- Verify schemas/ast/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast.utils');

ROLLBACK;
