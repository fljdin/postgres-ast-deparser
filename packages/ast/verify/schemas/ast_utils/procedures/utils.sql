-- Verify schemas/ast_utils/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast_utils.utils');

ROLLBACK;
