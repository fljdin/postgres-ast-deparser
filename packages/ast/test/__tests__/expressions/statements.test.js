import { getConnections } from '../../utils';
import { inserts } from './__fixtures__/inserts';
let db, teardown;
const objs = {
  tables: {}
};

beforeAll(async () => {
  ({ db, teardown } = await getConnections());
  await db.begin();
  await db.savepoint();
});
afterAll(async () => {
  try {
    //try catch here allows us to see the sql parsing issues!
    await db.rollback();
    await db.commit();
    await teardown();
  } catch (e) {}
});

it('func_call', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse( ast.func_call( 'dan' ));
  `);
  expect(result).toMatchSnapshot();
});

it('a_expr', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse( ast.a_expr(0, ast.string('a'), '=', ast.string('b')) );
  `);
  expect(result).toMatchSnapshot();
});

it('type_name', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse( ast.type_name( to_jsonb(ARRAY[ast.string('text')]) ) );
  `);
  expect(result).toMatchSnapshot();
});

it('type_cast', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse( ast.type_cast(ast.a_const(ast.null()), ast.type_name( to_jsonb(ARRAY[ast.string('text')]), true )) );
  `);
  expect(result).toMatchSnapshot();
});

it('raw_stmt', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.raw_stmt(
    ast.type_cast(ast.a_const(ast.null()), ast.type_name( to_jsonb(ARRAY[ast.string('text')]), true ))
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('rule_stmt', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.rule_stmt(
    'my-rule',
    3,
    ast.string('<relation placeholder>'),
    ast.string('<whereClause placeholder>'),
    to_jsonb(ARRAY[ ast.string('<actions placeholder>') ])
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('a_expr', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.a_expr(
      0,
      ast.string('<lexpr placeholder>'),
      '=',
      ast.string('<rexpr placeholder>')
    )
);
  `);
  expect(result).toMatchSnapshot();
});

it('a_expr 0-4', async () => {
  for (let i = 0; i < 5; i++) {
    const [{ deparse: result }] = await db.any(
      `
select deparser.deparse(
  ast.a_expr(
      $1,
      ast.string('<lexpr placeholder>'),
      '=',
      ast.string('<rexpr placeholder>')
    )
);
  `,
      [i]
    );
    expect(result).toMatchSnapshot();
  }
});

it('a_expr 0-4 name', async () => {
  for (let i = 0; i < 5; i++) {
    const [{ deparse: result }] = await db.any(
      `
select deparser.deparse(
  ast.a_expr(
      $1,
      ast.string('<lexpr placeholder>'),
      to_jsonb(ARRAY[ast.string('<name1 placeholder>'),ast.string('<name2 placeholder>')]),
      ast.string('<rexpr placeholder>')
    )
);
  `,
      [i]
    );
    expect(result).toMatchSnapshot();
  }
});

it('a_expr 5 ', async () => {
  for (let i = 5; i < 6; i++) {
    const [{ deparse: result }] = await db.any(
      `
select deparser.deparse(
  ast.a_expr(
      $1,
      ast.string('<lexpr placeholder>'),
      '=',
      to_jsonb(ARRAY[ast.string('<rexpr placeholder>')])
    )
);
  `,
      [i]
    );
    expect(result).toMatchSnapshot();
  }
});

it('a_expr 6-7 ', async () => {
  for (let i = 6; i < 8; i++) {
    const [{ deparse: result }] = await db.any(
      `
select deparser.deparse(
  ast.a_expr(
      $1,
      ast.string('<lexpr placeholder>'),
      '=',
      to_jsonb(ARRAY[ast.string('<rexpr placeholder>')])
    )
);
  `,
      [i]
    );
    expect(result).toMatchSnapshot();
  }
});

it('a_expr 8-9', async () => {
  for (let i = 8; i < 10; i++) {
    const [{ deparse: result }] = await db.any(
      `
select deparser.deparse(
  ast.a_expr(
      $1,
      ast.string('<lexpr placeholder>'),
      '=',
      ast.string('<rexpr placeholder>')
    )
);
  `,
      [i]
    );
    expect(result).toMatchSnapshot();
  }
});

it('a_expr 10', async () => {
  for (let i = 10; i < 11; i++) {
    const [{ deparse: result }] = await db.any(
      `
select deparser.deparse(
  ast.a_expr(
      $1,
      ast.string('<lexpr placeholder>'),
      '=',
      ast.func_call('func', to_jsonb(ARRAY[
        ast.string('arg1'),
        ast.string('arg2')
      ]))
    )
);
  `,
      [i]
    );
    expect(result).toMatchSnapshot();
  }
});

it('a_expr 11-12', async () => {
  for (let i = 11; i < 13; i++) {
    const [{ deparse: result }] = await db.any(
      `
select deparser.deparse(
  ast.a_expr(
      $1,
      ast.string('<lexpr placeholder>'),
      '=',
      to_jsonb(ARRAY[
        ast.string('arg1'),
        ast.string('arg2')
      ])
    )
);
  `,
      [i]
    );
    expect(result).toMatchSnapshot();
  }
});

it('alias', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.alias(
    'myrule',
    to_jsonb(ARRAY[ ast.string('namesplaceholder') ])
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('alias', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.alias(
    'myrule'
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('a_array_expr', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.a_array_expr(
    to_jsonb(ARRAY[ 
      ast.string('namesplaceholder1'),
      ast.string('namesplaceholder2')
    ])
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('a_const', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.a_const(
    'aconst'
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('a_const str', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.a_const(
    ast.string('astring')
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('a_const int', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.a_const(
    ast.integer(2)
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('a_indices', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.a_indices(
    ast.integer(2)
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('a_indices', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.a_indices(
    ast.integer(2),
    ast.integer(2)
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('a_indirection', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.a_indirection(
    ast.integer(2),
    to_jsonb(ARRAY[ 
      ast.string('namesplaceholder1'),
      ast.string('namesplaceholder2')
    ])
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('a_star', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.a_star( )
);
  `);
  expect(result).toMatchSnapshot();
});

it('bit_string', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.bit_string(
    'mystring'
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('bool_expr', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.bool_expr(
    0,
    to_jsonb(ARRAY[ 
      ast.string('namesplaceholder1'),
      ast.string('namesplaceholder2')
    ])
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('bool_expr', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.bool_expr(
    1,
    to_jsonb(ARRAY[ 
      ast.string('namesplaceholder1'),
      ast.string('namesplaceholder2')
    ])
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('bool_expr', async () => {
  const [{ deparse: result }] = await db.any(`
select deparser.deparse(
  ast.bool_expr(
    2,
    to_jsonb(ARRAY[ 
      ast.string('namesplaceholder1'),
      ast.string('namesplaceholder2')
    ])
  )
);
  `);
  expect(result).toMatchSnapshot();
});

it('boolean_test', async () => {
  for (let i = 0; i < 6; i++) {
    const [{ deparse: result }] = await db.any(
      `
select deparser.deparse(
  ast.boolean_test(
    $1,
    ast.string('<booltest placeholder>')
  )
);
  `,
      [i]
    );
    expect(result).toMatchSnapshot();
  }
});

it('case_expr', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.case_expr(
    ast.string('<arg placeholder>'),
    to_jsonb(ARRAY[ 
      ast.string('<arg1>'),
      ast.string('<arg2>')
    ]),
    ast.string('<defresult placeholder>')
  )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('case_expr', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.case_expr(
    ast.string('<arg placeholder>'),
    NULL,
    ast.string('<defresult placeholder>')
  )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('coalesce_expr', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.coalesce_expr(
    to_jsonb(ARRAY[ 
      ast.string('<arg1>'),
      ast.string('<arg2>')
    ])
    )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('collate_clause', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.collate_clause(
    ast.string('<arg1>'),
    to_jsonb(ARRAY[ 
      ast.string('<arg1>'),
      ast.string('<arg2>')
    ])
    )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('composite_type_stmt', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.composite_type_stmt(
    ast.string('<arg1>'),
    to_jsonb(ARRAY[ 
      ast.string('<arg1>'),
      ast.string('<arg2>')
    ])
    )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('rename_stmt', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.rename_stmt(
    6,
    6,
    ast.string('<arg1>'),
    'subname',
    'newname'
    )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('column_def', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.column_def(
    'mycol',
    ast.type_name( to_jsonb(ARRAY[ast.string('text')]) ),
    to_jsonb(ARRAY[ 
      ast.string('<arg1>'),
      ast.string('<arg2>')
    ]),
    ast.collate_clause(
      ast.string('<arg1>'),
      ast.string('<arg2>')
    ),
    ast.string('<arg3>')
   )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('column_def', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.column_def(
    'mycol',
    ast.type_name( to_jsonb(ARRAY[ast.string('text')]) ),
    to_jsonb(ARRAY[ 
      ast.string('<arg1>'),
      ast.string('<arg2>')
    ]),
    NULL,
    NULL
   )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('column_def', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.column_def(
    'my-col',
    ast.type_name( to_jsonb(ARRAY[ast.string('text')]) ),
    NULL,
    NULL,
    NULL
   )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('sql_value_function', async () => {
  for (const i of [0, 3, 10, 12]) {
    const [{ deparse: result }] = await db.any(
      `
select deparser.deparse(
  ast.sql_value_function(
    $1
   )
);
  `,
      [i]
    );
    expect(result).toMatchSnapshot();
  }
});

it('column_ref', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.column_ref(
    to_jsonb(ARRAY[ 
      ast.string('<arg1>'),
      ast.string('<arg2>')
    ])
   )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('column_ref', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.column_ref(
    to_jsonb(ARRAY[ 
      ast.string('arg1'),
      ast.string('arg2')
    ])
   )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('comment_stmt', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.comment_stmt(
    6,
    ast.string('<object>'),
    'my comment',
    to_jsonb(ARRAY[ 
      ast.string('arg1'),
      ast.string('arg2')
    ])
   )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('common_table_expr', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.common_table_expr(
    'ctename',
    ast.string('<ctequery>'),
    to_jsonb(ARRAY[ 
      ast.string('arg1'),
      ast.string('arg2')
    ])
   )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('def_elem', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.def_elem(
    'thing',
    ast.string('<arg>'),
    0
    )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('def_elem', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.def_elem(
    'transaction_isolation',
    ast.a_const(ast.string('mylevel')),
    0
    )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('do_stmt', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.do_stmt(
    'RUN MY CODEZ'
    )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('float', async () => {
  for (const i of [0.1234, 3234.234235234, 3.145, 234.2342432]) {
    const [{ deparse: result }] = await db.any(
      `
select deparser.deparse(
  ast.float(
    $1::float
  )
);
  `,
      [i]
    );
    expect(result).toMatchSnapshot();
  }
});

it('func_call', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.func_call(
    to_jsonb(ARRAY[ 
      ast.string('name1'),
      ast.string('name2')
    ]),
    to_jsonb(ARRAY[ 
      ast.string('arg1'),
      ast.string('arg2')
    ]),
    true,
    true
  )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('func_call', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.func_call(
    to_jsonb(ARRAY[ 
      ast.string('name1'),
      ast.string('name2')
    ]),
    to_jsonb(ARRAY[ 
      ast.string('arg1'),
      ast.string('arg2')
    ]),
    true,
    true,
    true,
    true,
    ast.string('arg2'),
    to_jsonb(ARRAY[ 
      ast.string('arg1'),
      ast.string('arg2')
    ]),
    ast.string('arg2')
  )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('func_call', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.func_call(
    to_jsonb(ARRAY[ 
      ast.string('name1'),
      ast.string('name2')
    ]),
    to_jsonb(ARRAY[ 
      ast.string('arg1'),
      ast.string('arg2')
    ]),
    false,
    false,
    false,
    false,
    ast.string('arg2'),
    to_jsonb(ARRAY[ 
      ast.string('arg1'),
      ast.string('arg2')
    ]),
    ast.string('arg2')
  )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('grouping_func', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.grouping_func(
    to_jsonb(ARRAY[ 
      ast.string('name1'),
      ast.string('name2')
    ])
  )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('grouping_set', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.grouping_set(
    0
  )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('grouping_set', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.grouping_set(
    2,
    to_jsonb(ARRAY[ 
      ast.string('name1'),
      ast.string('name2')
    ])
  )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('grouping_set', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.grouping_set(
    3,
    to_jsonb(ARRAY[ 
      ast.string('name1'),
      ast.string('name2')
    ])
  )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('grouping_set', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.grouping_set(
    4,
    to_jsonb(ARRAY[ 
      ast.string('name1'),
      ast.string('name2')
    ])
  )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('index_stmt', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.index_stmt(
    ast.string('<relation>')
  )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('index_stmt', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.index_stmt(
    ast.string('<relation>'),
    to_jsonb(ARRAY[ 
      ast.string('name1'),
      ast.string('name2')
    ])
  )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('index_stmt', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.index_stmt(
    ast.string('<relation>'),
    to_jsonb(ARRAY[ 
      ast.string('name1'),
      ast.string('name2')
    ])::jsonb,
    NULL::jsonb,
    true,
    'idxname'::text,
    true
  )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('index_elem', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.index_elem(
    'indexname',
    ast.string('<relation>')
  )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('insert_stmt', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.insert_stmt(
    ast.string('<relation>')
  )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('insert_stmt', async () => {
  const [{ deparse: result }] = await db.any(
    `
select deparser.deparse(
  ast.insert_stmt(
    ast.string('<relation>'),
    to_jsonb(ARRAY[ 
      ast.string('id'),
      ast.string('name')
    ])::jsonb
  )
);
  `
  );
  expect(result).toMatchSnapshot();
});

it('insert_stmt', async () => {
  for (const insert of inserts) {
    const [{ deparse: result }] = await db.any(
      `
select deparser.deparse(
  $1::jsonb
);
  `,
      [insert]
    );
    expect(result).toMatchSnapshot();
  }
});
