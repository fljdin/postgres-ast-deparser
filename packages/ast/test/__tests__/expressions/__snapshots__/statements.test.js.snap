// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a_array_expr 1`] = `"ARRAY[namesplaceholder1, namesplaceholder2]"`;

exports[`a_const 1`] = `"'aconst'"`;

exports[`a_const int 1`] = `"2"`;

exports[`a_const str 1`] = `"'astring'"`;

exports[`a_expr 0-4 1`] = `"(<lexpr placeholder> = <rexpr placeholder>)"`;

exports[`a_expr 0-4 2`] = `"<lexpr placeholder> = ANY( <rexpr placeholder> )"`;

exports[`a_expr 0-4 3`] = `"<lexpr placeholder> = ALL( <rexpr placeholder> )"`;

exports[`a_expr 0-4 4`] = `"<lexpr placeholder> IS DISTINCT FROM <rexpr placeholder>"`;

exports[`a_expr 0-4 5`] = `"NULLIF(<lexpr placeholder>, <rexpr placeholder>)"`;

exports[`a_expr 0-4 name 1`] = `"(<lexpr placeholder> OPERATOR(\\"<name1 placeholder>\\".<name2 placeholder>) <rexpr placeholder>)"`;

exports[`a_expr 0-4 name 2`] = `"<lexpr placeholder> <name1 placeholder> ANY( <rexpr placeholder> )"`;

exports[`a_expr 0-4 name 3`] = `"<lexpr placeholder> <name1 placeholder> ALL( <rexpr placeholder> )"`;

exports[`a_expr 0-4 name 4`] = `"<lexpr placeholder> IS DISTINCT FROM <rexpr placeholder>"`;

exports[`a_expr 0-4 name 5`] = `"NULLIF(<lexpr placeholder>, <rexpr placeholder>)"`;

exports[`a_expr 1`] = `"(a = b)"`;

exports[`a_expr 2`] = `"(<lexpr placeholder> = <rexpr placeholder>)"`;

exports[`a_expr 5  1`] = `"<lexpr placeholder> IS OF ( <rexpr placeholder> )"`;

exports[`a_expr 6-7  1`] = `"<lexpr placeholder> IN ( <rexpr placeholder> )"`;

exports[`a_expr 6-7  2`] = `"<lexpr placeholder> IN ( <rexpr placeholder> )"`;

exports[`a_expr 8-9 1`] = `"<lexpr placeholder> LIKE ( <rexpr placeholder> )"`;

exports[`a_expr 8-9 2`] = `"<lexpr placeholder> ILIKE ( <rexpr placeholder> )"`;

exports[`a_expr 10 1`] = `"<lexpr placeholder> SIMILAR TO arg1 ESCAPE arg2"`;

exports[`a_expr 11-12 1`] = `"<lexpr placeholder> BETWEEN arg1 AND arg2"`;

exports[`a_expr 11-12 2`] = `"<lexpr placeholder> NOT BETWEEN arg1 AND arg2"`;

exports[`a_indices 1`] = `"[2]"`;

exports[`a_indices 2`] = `"[2:2]"`;

exports[`a_indirection 1`] = `".namesplaceholder1.namesplaceholder2"`;

exports[`a_star 1`] = `"*"`;

exports[`alias 1`] = `"AS myrule (namesplaceholder)"`;

exports[`alias 2`] = `"AS myrule"`;

exports[`bit_string 1`] = `"m'ystring'"`;

exports[`bool_expr 1`] = `"(namesplaceholder1 AND namesplaceholder2)"`;

exports[`bool_expr 2`] = `"(namesplaceholder1 OR namesplaceholder2)"`;

exports[`bool_expr 3`] = `"NOT IN (namesplaceholder1)"`;

exports[`boolean_test 1`] = `"<booltest placeholder> IS TRUE"`;

exports[`boolean_test 2`] = `"<booltest placeholder> IS NOT TRUE"`;

exports[`boolean_test 3`] = `"<booltest placeholder> IS FALSE"`;

exports[`boolean_test 4`] = `"<booltest placeholder> IS NOT FALSE"`;

exports[`boolean_test 5`] = `"<booltest placeholder> IS UNKNOWN"`;

exports[`boolean_test 6`] = `"<booltest placeholder> IS NOT UNKNOWN"`;

exports[`case_expr 1`] = `"CASE <arg placeholder> <arg1> <arg2> ELSE <defresult placeholder> END"`;

exports[`case_expr 2`] = `"CASE <arg placeholder> ELSE <defresult placeholder> END"`;

exports[`coalesce_expr 1`] = `"COALESCE(<arg1>, <arg2>)"`;

exports[`collate_clause 1`] = `"<arg1> COLLATE \\"<arg1>\\", \\"<arg2>\\""`;

exports[`column_def 1`] = `"mycol text USING <arg3> <arg1> <arg2> COLLATE"`;

exports[`column_def 2`] = `"mycol text <arg1> <arg2>"`;

exports[`column_def 3`] = `"\\"my-col\\" text"`;

exports[`composite_type_stmt 1`] = `"CREATE TYPE <arg1> AS (<arg1>,<arg2>)"`;

exports[`func_call 1`] = `"dan ()"`;

exports[`raw_stmt 1`] = `"NULL::text[]"`;

exports[`rename_stmt 1`] = `"ALTER TABLE <arg1> RENAME COLUMN subname TO newname"`;

exports[`rule_stmt 1`] = `"CREATE RULE \\"my-rule\\" AS ON INSERT TO <relation placeholder> WHERE <whereClause placeholder> DO <actions placeholder>"`;

exports[`type_cast 1`] = `"NULL::text[]"`;

exports[`type_name 1`] = `"text"`;
